### DISCUSSION TITLE
useBlocker and React Hook Form

### DISCUSSION THREAD

**Discussion by ThaynarLima** [remix-run/react-router#11599] [Proposals]
Hello! 

I'm using useBlocker on my application. The scenario is the following:
1. I Have a form and if I try to leave the form with dirty values I get a modal explaining that I'll lose all data typed on the form.
2. User can confirm or cancel the navigation and then i use proceed for confirming and reset for canceling.
3. The problem is that when I click on save, and the user is saved I should be redirected to the list of items, but as react hook form reset function takes a bit to reset, the blocker is called even when it shouldn't, because it has the "old" isDirty true prop.

Is there a way useBlocker could have a dependency in which I would use my isDirty so it is updated when it gets called?my workaround for now is put a setTimeout around my navigation so it gets called after reset is done.

**Comment by david-crespo** [comment]
We are running into a similar issue. We don't use reset, so it's not `isDirty` we're concerned about. The problem is that `isSubmitSuccessful` is not turning to `true` before our `navigate` happens. It takes another 30ms or so after the submit callback finishes to see it turn true. This is true even when I manually put the `navigate()` call after `await handleSubmit(...)`.

https://github.com/oxidecomputer/console/blob/e8bf056cc1856f96793518a0b619026e91a883fa/app/components/form/FullPageForm.tsx#L58-L76

**Comment by with-key** [comment]
```useBlocker(formMethods.formState.isDirty)```

**Comment by User6531** [comment]
any updates?

**Comment by floooptimism** [comment]
I had this problem with RHF's reset as well. What I did was add another condition in my `useBlocker` condition that isn't managed by React. In my case, I used Jotai, but you could use anything, like storing it in a singleton object.  

#### **Setting up an external store with Jotai:**
```js
const blockerStore = createStore();
export const blockerActiveOverrideAtom = atom(true);
export const setIsBlockerActiveOverride = (val) => blockerStore.set(blockerActiveOverrideAtom, val);
export const getIsBlockerActiveOverride = () => blockerStore.get(blockerActiveOverrideAtom);
```

my useBlocker would look like this
```js
const blocker = useBlocker(() => isDirty && getIsBlockerActiveOverride());
```

So instead of resetting the form, you can just set `setIsBlockerActiveOverride ` to `false` and should navigate out without getting blocked.

**Comment by bycolour2** [comment]
Any updates?
Preferably using "natural" rhf & rrd methods

**Comment by nemmtor** [comment]
I was able to solve this problem by wrapping reset in a `flushSync`, hope it helps:
```tsx
  const submit: SubmitHandler<FormValues> = (values) => {
    flushSync(() => {
      formMethods.reset(values);
    });
    navigate('/baz');
  };

```

**Reply by Mario-Eis** [reply]
> I was able to solve this problem by wrapping reset in a `flushSync`, hope it helps:
> 
> ```tsx
>   const submit: SubmitHandler<FormValues> = (values) => {
>     flushSync(() => {
>       formMethods.reset(values);
>     });
>     navigate('/baz');
>   };
> ```

Why is this downvoted? this seems like the only reliable soluton so far.
